File: leetcode/201-300/269-alien-dictionary/bench/alien.c
// Benchmark workload for LeetCode #269 — Alien Dictionary (C mirror).
// Mirrors alien.kara algorithm-for-algorithm, including the flat corpus and the
// hoisted working structures (see that file for why).
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { long *d; long n, cap; } Vec;
static void vpush(Vec *v, long x) {
    if (v->n == v->cap) { v->cap = v->cap ? v->cap * 2 : 64; v->d = realloc(v->d, v->cap * sizeof(long)); }
    v->d[v->n++] = x;
}

int main(void) {
    long lists = 20000, rounds = 60, alpha = 8;

    Vec letters = {0}, wstart = {0}, wlen = {0}, lstart = {0}, lcount = {0};
    long state = 269269;

    for (long li = 0; li < lists; li++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long m = (state / 65536L) % 5 + 2;

        long rank[8];
        for (long z = 0; z < alpha; z++) rank[z] = z;
        for (long sh = alpha - 1; sh > 0; sh--) {
            state = (state * 1103515245L + 12345L) & 2147483647L;
            long j = (state / 65536L) % (sh + 1);
            long t = rank[sh]; rank[sh] = rank[j]; rank[j] = t;
        }

        long buf[64], st[8], ln[8], bn = 0;
        for (long wi = 0; wi < m; wi++) {
            state = (state * 1103515245L + 12345L) & 2147483647L;
            long L = (state / 65536L) % 4 + 1;
            st[wi] = bn; ln[wi] = L;
            for (long p = 0; p < L; p++) {
                state = (state * 1103515245L + 12345L) & 2147483647L;
                buf[bn++] = (state / 65536L) % alpha;
            }
        }

        for (long a = 1; a < m; a++) {
            long b = a;
            while (b > 0) {
                long s1 = st[b-1], n1 = ln[b-1], s2 = st[b], n2 = ln[b];
                long lim = n2 < n1 ? n2 : n1;
                long k = 0; int swap = 0, decided = 0;
                while (k < lim) {
                    if (buf[s1+k] != buf[s2+k]) {
                        if (rank[buf[s1+k]] > rank[buf[s2+k]]) swap = 1;
                        decided = 1; k = lim;
                    } else k++;
                }
                if (!decided && n1 > n2) swap = 1;
                if (swap) {
                    long ts = st[b-1]; st[b-1] = st[b]; st[b] = ts;
                    long tl = ln[b-1]; ln[b-1] = ln[b]; ln[b] = tl;
                    b--;
                } else b = 0;
            }
        }

        state = (state * 1103515245L + 12345L) & 2147483647L;
        if ((state / 65536L) % 2 == 0 && m >= 2) {
            state = (state * 1103515245L + 12345L) & 2147483647L;
            long at = (state / 65536L) % (m - 1);
            long ts = st[at]; st[at] = st[at+1]; st[at+1] = ts;
            long tl = ln[at]; ln[at] = ln[at+1]; ln[at+1] = tl;
        }

        vpush(&lstart, wstart.n);
        vpush(&lcount, m);
        for (long q = 0; q < m; q++) {
            vpush(&wstart, letters.n);
            vpush(&wlen, ln[q]);
            for (long r = 0; r < ln[q]; r++) vpush(&letters, buf[st[q] + r]);
        }
    }

    char adj[676];
    long indeg[26];
    char present[26], done[26];

    long sink = 0;
    for (long r0 = 0; r0 < rounds; r0++) {
        for (long idx = 0; idx < lists; idx++) {
            long base = lstart.d[idx], n = lcount.d[idx];

            for (long c = 0; c < 26; c++) { indeg[c] = 0; present[c] = 0; done[c] = 0; }
            memset(adj, 0, sizeof adj);

            for (long w = 0; w < n; w++) {
                long s = wstart.d[base + w], L = wlen.d[base + w];
                for (long p = 0; p < L; p++) present[letters.d[s + p]] = 1;
            }

            int bad = 0;
            long p2 = 0;
            while (p2 + 1 < n) {
                long s1 = wstart.d[base+p2], n1 = wlen.d[base+p2];
                long s2 = wstart.d[base+p2+1], n2 = wlen.d[base+p2+1];
                long lim = n2 < n1 ? n2 : n1;
                long k = 0; int found = 0;
                while (k < lim) {
                    long x = letters.d[s1+k], y = letters.d[s2+k];
                    if (x != y) {
                        if (!adj[x*26+y]) { adj[x*26+y] = 1; indeg[y]++; }
                        found = 1; k = lim;
                    } else k++;
                }
                if (!found && n1 > n2) { bad = 1; p2 = n; } else p2++;
            }

            long acc = 0;
            if (!bad) {
                long remaining = 0;
                for (long d = 0; d < 26; d++) if (present[d]) remaining++;
                long placed = 0;
                while (placed < remaining) {
                    long pick = -1;
                    for (long s3 = 0; s3 < 26; ) {
                        if (present[s3] && !done[s3] && indeg[s3] == 0) { pick = s3; s3 = 26; }
                        else s3++;
                    }
                    if (pick < 0) { acc = 0; placed = remaining; }
                    else {
                        done[pick] = 1;
                        acc = (acc * 31 + pick + 1) % 1000000007L;
                        placed++;
                        for (long t = 0; t < 26; t++) if (adj[pick*26+t]) indeg[t]--;
                    }
                }
            }
            sink = (sink * 131 + acc) % 1000000007L;
        }
    }
    printf("%ld\n", sink);
    return 0;
}

File: leetcode/201-300/269-alien-dictionary/bench/alien.rs
// Benchmark workload for LeetCode #269 — Alien Dictionary (Rust mirror).
// Mirrors alien.kara algorithm-for-algorithm, including the flat corpus and the
// hoisted working structures (see that file for why).

fn main() {
    let lists: i64 = 20000;
    let rounds: i64 = 60;
    let alpha: i64 = 8;

    let mut letters: Vec<i64> = Vec::new();
    let mut wstart: Vec<i64> = Vec::new();
    let mut wlen: Vec<i64> = Vec::new();
    let mut lstart: Vec<i64> = Vec::new();
    let mut lcount: Vec<i64> = Vec::new();

    let mut state: i64 = 269269;
    for _ in 0..lists {
        state = (state * 1103515245 + 12345) & 2147483647;
        let m = (state / 65536) % 5 + 2;

        let mut rank: Vec<i64> = (0..alpha).collect();
        let mut sh = alpha - 1;
        while sh > 0 {
            state = (state * 1103515245 + 12345) & 2147483647;
            let j = (state / 65536) % (sh + 1);
            rank.swap(sh as usize, j as usize);
            sh -= 1;
        }

        let mut buf: Vec<i64> = Vec::new();
        let mut st: Vec<i64> = Vec::new();
        let mut ln: Vec<i64> = Vec::new();
        for _ in 0..m {
            state = (state * 1103515245 + 12345) & 2147483647;
            let l = (state / 65536) % 4 + 1;
            st.push(buf.len() as i64);
            ln.push(l);
            for _ in 0..l {
                state = (state * 1103515245 + 12345) & 2147483647;
                buf.push((state / 65536) % alpha);
            }
        }

        let mut a = 1;
        while a < m {
            let mut b = a;
            while b > 0 {
                let (s1, n1) = (st[(b - 1) as usize], ln[(b - 1) as usize]);
                let (s2, n2) = (st[b as usize], ln[b as usize]);
                let lim = if n2 < n1 { n2 } else { n1 };
                let mut k = 0;
                let mut swap = false;
                let mut decided = false;
                while k < lim {
                    if buf[(s1 + k) as usize] != buf[(s2 + k) as usize] {
                        if rank[buf[(s1 + k) as usize] as usize] > rank[buf[(s2 + k) as usize] as usize] {
                            swap = true;
                        }
                        decided = true;
                        k = lim;
                    } else {
                        k += 1;
                    }
                }
                if !decided && n1 > n2 { swap = true; }
                if swap {
                    st.swap((b - 1) as usize, b as usize);
                    ln.swap((b - 1) as usize, b as usize);
                    b -= 1;
                } else {
                    b = 0;
                }
            }
            a += 1;
        }

        state = (state * 1103515245 + 12345) & 2147483647;
        if (state / 65536) % 2 == 0 && m >= 2 {
            state = (state * 1103515245 + 12345) & 2147483647;
            let at = (state / 65536) % (m - 1);
            st.swap(at as usize, (at + 1) as usize);
            ln.swap(at as usize, (at + 1) as usize);
        }

        lstart.push(wstart.len() as i64);
        lcount.push(m);
        for q in 0..m {
            wstart.push(letters.len() as i64);
            wlen.push(ln[q as usize]);
            for r in 0..ln[q as usize] {
                letters.push(buf[(st[q as usize] + r) as usize]);
            }
        }
    }

    let mut adj = vec![false; 676];
    let mut indeg = vec![0i64; 26];
    let mut present = vec![false; 26];
    let mut done = vec![false; 26];

    let mut sink: i64 = 0;
    for _ in 0..rounds {
        for idx in 0..lists {
            let base = lstart[idx as usize];
            let n = lcount[idx as usize];

            for c in 0..26 {
                indeg[c] = 0;
                present[c] = false;
                done[c] = false;
            }
            for e in 0..676 { adj[e] = false; }

            for w in 0..n {
                let s = wstart[(base + w) as usize];
                let l = wlen[(base + w) as usize];
                for p in 0..l {
                    present[letters[(s + p) as usize] as usize] = true;
                }
            }

            let mut bad = false;
            let mut p2 = 0i64;
            while p2 + 1 < n {
                let (s1, n1) = (wstart[(base + p2) as usize], wlen[(base + p2) as usize]);
                let (s2, n2) = (wstart[(base + p2 + 1) as usize], wlen[(base + p2 + 1) as usize]);
                let lim = if n2 < n1 { n2 } else { n1 };
                let mut k = 0;
                let mut found = false;
                while k < lim {
                    let x = letters[(s1 + k) as usize];
                    let y = letters[(s2 + k) as usize];
                    if x != y {
                        if !adj[(x * 26 + y) as usize] {
                            adj[(x * 26 + y) as usize] = true;
                            indeg[y as usize] += 1;
                        }
                        found = true;
                        k = lim;
                    } else {
                        k += 1;
                    }
                }
                if !found && n1 > n2 { bad = true; p2 = n; } else { p2 += 1; }
            }

            let mut acc: i64 = 0;
            if !bad {
                let mut remaining = 0i64;
                for d in 0..26 { if present[d] { remaining += 1; } }
                let mut placed = 0i64;
                while placed < remaining {
                    let mut pick: i64 = -1;
                    let mut s3 = 0usize;
                    while s3 < 26 {
                        if present[s3] && !done[s3] && indeg[s3] == 0 {
                            pick = s3 as i64;
                            s3 = 26;
                        } else {
                            s3 += 1;
                        }
                    }
                    if pick < 0 {
                        acc = 0;
                        placed = remaining;
                    } else {
                        done[pick as usize] = true;
                        acc = (acc * 31 + pick + 1) % 1000000007;
                        placed += 1;
                        for t in 0..26 {
                            if adj[(pick * 26 + t as i64) as usize] { indeg[t] -= 1; }
                        }
                    }
                }
            }
            sink = (sink * 131 + acc) % 1000000007;
        }
    }

    println!("{}", sink);
}
